File: HackerRank/search/task.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <utility>
using namespace std;

vector<bool> used;
int mid(pair<int, int> x)
{
	return (x.first + x.second) >> 1;
}

class Task
{
	public:
		int idx, d, m;
		
		bool operator<(const Task& t) const
		{
			if(this->d!=t.d)
				return this->d < t.d;
			if(this->m!=t.m)
				return this->m > t.m;
			return this->idx < t.idx;
		}
};

template<class T>
class Node
{
	private:
		pair<int, int> interval;
		
		void push()
		{
			left->val += this->add;
			left->add += this->add;
			right->val += this->add;
			right->add += this->add;
			this->add = 0;
		}
	
	public:
		T val, add, idx;
		
		Node()
		{
			this->add=0;
			this->left= this->right = NULL;
		}
		
		void build(int s, int e)
		{
			//cout << s << ' ' << e << endl;
			this->add=0;
			this->val=0;
			this->idx=s;
			interval.first=s, interval.second=e;
			if(s==e)
				return;
			
			this->left=new Node<T>();
			this->right=new Node<T>();
			int m = mid(this->interval);
			this->left->build(s, m);
			this->right->build(m+1, e);
		}
		
		void update(int s, int e, int v)
		{
			if(s<=interval.first && interval.second<=e)
			{
				this->val += v;
				this->add += v;
				return;
			}
			if(this->add)
				this->push();
			int m = mid(this->interval);
			if(s<=m)
				left->update(s, e, v);
			if(m<e)
				right->update(s, e, v);
			
			Node mn = max(*left, *right);
			this->val = mn.val;
			this->idx = mn.idx;
		}
		
		Node query(int s, int e)
		{
			if(s<=interval.first && interval.second<=e)
				return *this;
			
			if(this->add)
				this->push();
			int m = mid(this->interval);
			
			if(e<=m)
				return left->query(s, e);
			else if(s>m)
				return right->query(s, e);
			else
				return max(left->query(s, m), right->query(m+1, e));
		}
		
		bool operator<(const Node& x) const
		{
			if(!used[this->idx])
				return true;
			if(!used[x.idx])
				return false;
			return this->val < x.val;
		}
		Node<T> *left, *right;
};

template<class T>
class BIT
{
	private:
		vector<T> A;
	
	public:
		BIT(int n)
		{
			A.resize(n);
		}
		
		T get(int idx)
		{
			T res=0;
			while(idx)
			{
				res+=A[idx];
				idx -= idx&-idx;
			}
			return res;
		}
		
		void update(int idx, T val=1)
		{
			while(idx<=A.size())
			{
				A[idx] += val;
				idx += idx&-idx;
			}
		}
};

int main()
{
	int T;
	cin >> T;
	vector<Task> task(T);
	used.resize(T+10);
	for(int i=0;i<T;i++)
		cin >> task[i].d >> task[i].m, task[i].idx=i;
	
	sort(task.begin(), task.end());
	vector<int> rank(T);
	for(int i=0;i<T;i++)
		rank[task[i].idx] = i;
	
	int ans=0;
	BIT<int> sum(T+10);
	Node<int> segT;
	segT.build(1, T);
	
	for(int i=0;i<T;i++)
	{
		int idx=rank[i]+1, res;
		
		sum.update(idx, task[idx-1].m);
		used[idx] = true;
		
		segT.update(idx, idx, -task[idx-1].d);
		segT.update(idx, T, task[idx-1].m);
		
		//cout << sum.get(idx+1) << ' ' << task[idx].d << endl;
		
		res = max(0, sum.get(idx) - task[idx-1].d);
		if(idx+1 <= T)
		{
			//cout << "lan:" << idx+1 << endl;
			Node<int> r = segT.query(idx+1, T);
			if(used[r.idx])
			{
				res = max(res, sum.get(r.idx) - task[r.idx-1].d);
				//cout << r.idx << ' ' << sum.get(r.idx) << ' ' << task[r.idx-1].d << endl;
			}
		}
		
		ans = max(res, ans);
		cout << ans << endl;
	}
}

File: HackerRank/search/task.py
import sys
import math
 
stdin = sys.stdin
 
Mid= lambda x : ((x[0] + x[1])>>1)
Lowbit = lambda x : (x&(-x))
 
used = {}
 
class Task:
	def __init__(self , _d , _m , _index):
		self.d = int(_d)
		self.m = int(_m)
		self.index = int(_index)
	def __lt__(self , other):
		if (self.d == other.d):
			return self.m > other.m
		else:
			return self.d < other.d
	def __repr__(self):
		return str((self.d , self.m))
class Node:
	def __init__(self):
		self.val = 0
		self.idx = 0
		self.add = 0
		self.cover = (0 , 0)
		self.lChild = None
		self.rChild = None
	def __lt__(self , other):
		if (self.idx not in used):
			return True
		if (other.idx not in used):
			return False
		return self.val < other.val
	def __repr__(self):
		return str(self.val)
	def build(self , cover):
		self.cover = cover
		self.idx = cover[0]
		if (cover[0] == cover[1]):
			return
		mid = Mid(cover)
		self.lChild = Node()
		self.rChild = Node()
		self.lChild.build( (cover[0] , mid) )
		self.rChild.build( (mid + 1 , cover[1]) )
	def addDown(self):
 
		self.lChild.addValue(self.add)
		self.rChild.addValue(self.add)
		self.add = 0
 
	def update(self , cover, add):
		if (cover[0] <= self.cover[0] and cover[1] >= self.cover[1]):
			self.val += add
			self.add += add
			return
		if (self.add):
			self.addDown()
		mid = Mid(self.cover)
		if (cover[0] <= mid):
			self.lChild.update(cover , add)
		if (cover[1] > mid):
			self.rChild.update(cover , add)
 
		ret = max( self.lChild , self.rChild )
		self.val = ret.val
		self.idx = ret.idx
 
	def query(self , cover):
		if (cover[0] == self.cover[0] and cover[1] == self.cover[1]):
			return self
		if (self.add):
			self.addDown()
		mid = Mid(self.cover)
		if (cover[1] <= mid):
			return self.lChild.query(cover)
		elif (cover[0] > mid):
			return self.rChild.query(cover)
		else:
			vl = self.lChild.query( (cover[0] , mid) )
			vr = self.rChild.query( (mid + 1 , cover[1]) )
			return max(vl , vr)
 
class BinIdxTree:
	def __init__(self , size):
		self.S = [0 for i in range(size+10)]
	def add(self , x , c):
 
		while (x <= n):
			self.S[x] += c
			x += Lowbit(x)
	def getsum(self , x):
		res = 0
		while (x > 0):
			res += self.S[x]
			x -= Lowbit(x)
		return res
 
 
line = stdin.readline()
n = int(line)
 
segTreeRoot = Node()
segTreeRoot.build((1 , n))
bTree = BinIdxTree(n)
 
taskList = []
rank = [0 for i in range(n+1)]
 
for i in range(n):
	line = stdin.readline()
	[d , m] = [int(x) for x in line.split()]
	t = Task(d , m , i)
	taskList.append(t)
 
taskList.sort()
 
for i in range(n):
	rank[ taskList[i].index ] = i
 
ans = 0
for i in range(n):
	p = rank[i] + 1
	bTree.add(p , taskList[p-1].m)
	used[p] = True
	segTreeRoot.update((p , p) ,  - taskList[p-1].d)
	segTreeRoot.update((p , n) , taskList[p-1].m)
	res = bTree.getsum(p) - taskList[p-1].d
	res = 0 if res < 0 else res
	if (p + 1 <= n):
		resNode = segTreeRoot.query((p + 1 , n))
		p = resNode.idx
		if (p in used):
			print p,bTree.getsum(p), taskList[p-1].d
			s =bTree.getsum(p) - taskList[p-1].d
			res = s if s > res else res
	ans = res if res > ans else ans
	print ans

